packages/roosterjs-editor-core/lib/coreApi/transformColor.ts

```diff
@@ -1,128 +1,130 @@
-
-    if (safeInstanceOf(rootNode, 'HTMLElement')) {
-        if (includeSelf) {
-            result.push(rootNode);
-        }
-        const allChildren = rootNode.getElementsByTagName('*');
-        arrayPush(result, toArray(allChildren));
-    } else if (safeInstanceOf(rootNode, 'DocumentFragment')) {
-        const allChildren = rootNode.querySelectorAll('*');
-        arrayPush(result, toArray(allChildren));
-    }
-
-    return result;
-}
+import { arrayPush, getComputedStyles, safeInstanceOf, toArray } from 'roosterjs-editor-dom';
+import {
+    ColorTransformDirection,
+    DarkModeDatasetNames,
+    EditorCore,
+    TransformColor,
+} from 'roosterjs-editor-types';
+
+const enum ColorAttributeEnum {
+    CssColor = 0,
+    HtmlColor = 1,
+    CssDataSet = 2,
+    HtmlDataSet = 3,
+}
+
+const ColorAttributeName: { [key in ColorAttributeEnum]: string }[] = [
+    {
+        [ColorAttributeEnum.CssColor]: 'color',
+        [ColorAttributeEnum.HtmlColor]: 'color',
+        [ColorAttributeEnum.CssDataSet]: DarkModeDatasetNames.OriginalStyleColor,
+        [ColorAttributeEnum.HtmlDataSet]: DarkModeDatasetNames.OriginalAttributeColor,
+    },
+    {
+        [ColorAttributeEnum.CssColor]: 'background-color',
+        [ColorAttributeEnum.HtmlColor]: 'bgcolor',
+        [ColorAttributeEnum.CssDataSet]: DarkModeDatasetNames.OriginalStyleBackgroundColor,
+        [ColorAttributeEnum.HtmlDataSet]: DarkModeDatasetNames.OriginalAttributeBackgroundColor,
+    },
+];
+
+/**
+ * @internal
+ * Edit and transform color of elements between light mode and dark mode
+ * @param core The EditorCore object
+ * @param rootNode The root HTML elements to transform
+ * @param includeSelf True to transform the root node as well, otherwise false
+ * @param callback The callback function to invoke before do color transformation
+ * @param direction To specify the transform direction, light to dark, or dark to light
+ */
+export const transformColor: TransformColor = (
+    core: EditorCore,
+    rootNode: Node,
+    includeSelf: boolean,
+    callback: () => void,
+    direction: ColorTransformDirection
+) => {
+    const elementsToTransform = core.lifecycle.isDarkMode ? getAll(rootNode, includeSelf) : [];
+    const transformFunction =
+        direction == ColorTransformDirection.DarkToLight
+            ? transformToLightMode
+            : core.lifecycle.onExternalContentTransform ||
+              ((element: HTMLElement) => transformToDarkMode(element, core.lifecycle.getDarkColor));
+
+    callback?.();
+
+    elementsToTransform.forEach(
+        element => element?.dataset && element.style && transformFunction(element)
+    );
+};
+
+function transformToLightMode(element: HTMLElement) {
+    ColorAttributeName.forEach(names => {
+        // Reset color styles based on the content of the ogsc/ogsb data element.
+        // If those data properties are empty or do not exist, ignore them.
+        if (element.dataset[names[ColorAttributeEnum.CssDataSet]]) {

```

Comment:
do you still need this change?

Reply:
Yes, because we need to set the color for light mode, if we remove this code, when w switch back to light mode, the dark to light transform will not happen.

Reply:
But the change is to check if there is data-ogsc, and not run if it doesn't have. That means after this change there will be fewer chance to run this code (only when data-ogsc exist).